crates/chia-consensus/src/gen/condition_tools.rs

```diff
@@ -0,0 +1,86 @@
+use crate::consensus_constants::ConsensusConstants;
+use crate::gen::opcodes::{
+    ConditionOpcode, AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT,
+    AGG_SIG_PARENT_PUZZLE, AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::OwnedSpend;
+use chia_protocol::Bytes;
+use chia_protocol::Coin;
+
+pub fn make_aggsig_final_message(
+    opcode: ConditionOpcode,
+    msg: &[u8],
+    spend: &OwnedSpend,
+    constants: &ConsensusConstants,
+) -> Vec<u8> {
+    let mut result = msg.to_vec();
+    result.extend(match opcode {
+        AGG_SIG_PARENT => [
+            spend.parent_id.as_slice(),
+            constants.agg_sig_parent_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PUZZLE => [
+            spend.puzzle_hash.as_slice(),
+            constants.agg_sig_puzzle_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_AMOUNT => [
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PUZZLE_AMOUNT => [
+            spend.puzzle_hash.as_slice(),
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_puzzle_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PARENT_AMOUNT => [
+            spend.parent_id.as_slice(),
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_parent_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PARENT_PUZZLE => [
+            spend.parent_id.as_slice(),
+            spend.puzzle_hash.as_slice(),
+            constants.agg_sig_parent_puzzle_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_ME => {
+            let coin: Coin = Coin::new(spend.parent_id, spend.puzzle_hash, spend.coin_amount);
+            [
+                coin.coin_id().as_slice(),
+                constants.agg_sig_me_additional_data.as_slice(),
+            ]
+            .concat()
```

Comment:
it would be a bit more efficient to `.extend()` the vector twice, rather than first concatenating the buffers, and then extend by the result. You get one more allocation the way you do it right now.

In fact, even better would be to allocate `Vec<u8>::with_capacity()` and specify the exact number of bytes, then `.extend()` the `msg` and the two or three parameters. Then you'd only have a single heap allocation.